src/openrct2/actions/StaffSetPatrolAreaAction.hpp

```diff
@@ -0,0 +1,99 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2019 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#include "../interface/Window.h"
+#include "../peep/Peep.h"
+#include "../peep/Staff.h"
+#include "../world/Sprite.h"
+#include "GameAction.h"
+
+DEFINE_GAME_ACTION(StaffSetPatrolAreaAction, GAME_COMMAND_SET_STAFF_PATROL, GameActionResult)
+{
+private:
+    uint16_t _spriteId;
+    CoordsXY _loc;
+
+public:
+    StaffSetPatrolAreaAction()
+    {
+    }
+    StaffSetPatrolAreaAction(uint16_t spriteId, CoordsXY loc)
+        : _spriteId(spriteId)
+        , _loc(loc)
+    {
+    }
+
+    uint16_t GetActionFlags() const override
+    {
+        return GameAction::GetActionFlags() | GA_FLAGS::ALLOW_WHILE_PAUSED;
+    }
+
+    void Serialise(DataSerialiser & stream) override
+    {
+        GameAction::Serialise(stream);
+        stream << DS_TAG(_spriteId) << DS_TAG(_loc);
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        if (_spriteId >= MAX_SPRITES)
+        {
+            log_error("Invalid spriteId.");
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        auto peep = GET_PEEP(_spriteId);
+        if (peep == nullptr || peep->sprite_identifier != SPRITE_IDENTIFIER_PEEP || peep->type != PEEP_TYPE_STAFF)
+        {
+            log_error("Invalid spriteId.");
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        return MakeResult();
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        auto peep = GET_PEEP(_spriteId);
+        if (peep == nullptr || peep->sprite_identifier != SPRITE_IDENTIFIER_PEEP || peep->type != PEEP_TYPE_STAFF)
+        {
+            log_error("Invalid spriteId.");
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        int32_t patrolOffset = peep->staff_id * STAFF_PATROL_AREA_SIZE;
+
+        staff_toggle_patrol_area(peep->staff_id, _loc.x, _loc.y);
+
+        int32_t ispatrolling = 0;
+        for (int32_t i = 0; i < 128; i++)
+        {
+            ispatrolling |= gStaffPatrolAreas[patrolOffset + i];
```

Comment:
This seems somewhat odd given that it iterates 128 times, where only 32 bits can be possibly set. I'm not sure if its ever the case but it might make sense to break if ispatrolling == std::numeric_limits<uint32_t>::max()

Reply:
I decided this code was inefficient and ugly. I've replaced isPatrolling with a bool since none of this actually matters.